cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -445,21 +479,58 @@ fn insert_common_prologue(
             // Also, the size of a return address, implicitly pushed by a x86 `call` instruction,
             // also should be accounted for.
             // TODO: Check if the function body actually contains a `call` instruction.
-            let word_size = isa.pointer_bytes();
             let total_stack_size = (csrs.iter(GPR).len() + 1 + 1) as i64 * word_size as i64;
 
             insert_stack_check(pos, total_stack_size, stack_limit_arg);
         }
     }
 
+    pos.func.frame_layout.initial = vec![
+        FrameLayoutChange::CallFrameAddressAt {
+            reg: cfa_state.reg,
+            offset: cfa_state.offset,
+        },
+        FrameLayoutChange::RaAt {
+            cfa_offset: cfa_state.position,
+        },
+    ]
+    .into_boxed_slice();
+
     // Append param to entry EBB
     let ebb = pos.current_ebb().expect("missing ebb under cursor");
     let fp = pos.func.dfg.append_ebb_param(ebb, reg_type);
     pos.func.locations[fp] = ir::ValueLoc::Reg(RU::rbp as RegUnit);
 
-    pos.ins().x86_push(fp);
-    pos.ins()
+    let push_fp_inst = pos.ins().x86_push(fp);
+    let word_size = word_size as isize;
+    cfa_state.position -= word_size;
+    cfa_state.offset += word_size;
```

Comment:
Are we sure no compiler pass won't add more instructions in between? (in particular thinking about register allocation which could insert copies/fills, or constant pools being generated, whenever we have them?)

Reply:
the `prologue_epilogue` runs after `regalloc` but before `shrink_instructions`, `relax_branches` -- I hope the latter does not insert anything to change stack layout.

Reply:
Right. There should be @julian-seward1 's late stage optimizations happening after that (and they can remove instructions), so we should be careful to put this somewhere the data flow graph can't change (i.e. at the very end).